### PR TITLE
Avoid value type inline allocations

### PR DESCRIPTION
**Pull request by hzongaro**
For the prototype implementation of value types, special handling of value type classes is required in a few cases:

- If the `new` bytecode is applied to a value type class, an `InstantiationError` must be thrown;
- If an identity (that is, non-value type) class has a value type field, that field should be instantiated with the default value of the value type when an instance of the identity class is create with the `new` bytecode; and
- If an array with value type elements is created with `anewarray` or with `multianewarray`, the elements of value type are initialized with the default value for the value type.

These changes avoid inline allocation and stack allocation in such situations, as the code for inline allocation is not yet able to handle the special requirements.

### CODE REVIEW

**Comment by hzongaro** [comment]
Andrew @andrewcraik, may I ask you to review the JIT changes in this pull request?
Dan @DanHeidinga, may I ask you to review the helper change in commit https://github.com/eclipse/openj9/pull/9139/commits/e91dcf5de46b3fb00e676d5474e337e7d3e411ab?

File: runtime/compiler/optimizer/EscapeAnalysis.cpp [hunk 1]
```diff
@@ -1076,6 +1076,23 @@ int32_t TR_EscapeAnalysis::performAnalysisOnce()
...
+         // with their default values.  EA is not yet set up to perform such iniitialization,
+         // so remove the candidate from consideration.
+         if (candidate->_kind == TR::New)
+            {
+            TR_OpaqueClassBlock *clazz = (TR_OpaqueClassBlock *)candidate->_node->getFirstChild()->getSymbol()->getStaticSymbol()->getStaticAddress();
+
+            if (!TR::Compiler->cls.isZeroInitializable(clazz))
+               {
+               if (trace())
+                  traceMsg(comp(), "   Fail [%p] because the candidate is not zero initializable (that is, it has a field that is an unflattened value type)\n", candidate->_node);
```
Comment (Leonardo2718) [inline runtime/compiler/optimizer/EscapeAnalysis.cpp:1089]: Just to clarify, "unflattened value type" means a value type that is not the reference project of the type?

File: runtime/compiler/compile/J9Compilation.cpp [hunk 2]
```diff
@@ -617,6 +617,8 @@ J9::Compilation::canAllocateInline(TR::Node* node, TR_OpaqueClassBlock* &classIn
 
    bool generateArraylets = self()->generateArraylets();
 
+   bool areValueTypesEnabled = TR::Compiler->om.areValueTypesEnabled();
```
Comment (Leonardo2718) [inline runtime/compiler/compile/J9Compilation.cpp:620]: Can I suggest making this a `const`?

File: runtime/compiler/optimizer/EscapeAnalysis.cpp [hunk 3]
```diff
@@ -1169,6 +1187,16 @@ int32_t TR_EscapeAnalysis::performAnalysisOnce()
...
                _candidates.remove(candidate);
                }
+            else
+               {
+               TR_OpaqueClassBlock *clazz = (TR_OpaqueClassBlock*)classNode->getSymbol()->castToStaticSymbol()->getStaticAddress();
+
+               if (TR::Compiler->cls.isValueTypeClass(clazz))
+                  {
+                  if (trace())
+                     traceMsg(comp(), "   Fail [%p] because array has value type elements\n", candidate->_node);
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/EscapeAnalysis.cpp:1197]: How does this fail happen - the trace says there is a failure, but the candidate isn't removed?

File: runtime/compiler/compile/J9Compilation.cpp [hunk 4]
```diff
@@ -665,11 +667,12 @@ J9::Compilation::canAllocateInline(TR::Node* node, TR_OpaqueClassBlock* &classIn
...
       classRef      = node->getSecondChild();
 
-      // In the case of dynamic array allocation, return 0 indicating variable dynamic array allocation
+      // In the case of dynamic array allocation, return 0 indicating variable dynamic array allocation,
+      // unless value types are enabled, in which case return -1 to prevent inline allocation
       if (classRef->getOpCodeValue() != TR::loadaddr)
          {
          classInfo = NULL;
-         return 0;
+         return areValueTypesEnabled ? -1 : 0;
```
Comment (andrewcraik) [inline runtime/compiler/compile/J9Compilation.cpp:675]: This is a pretty big thing since the dynamic allocations won't work - can we have a debug counter and a trace statement? Ultimately we would want to check at runtime if we could do the inline allocation otherwise call out, but this is fine as a first step for functional correctness.

File: runtime/compiler/optimizer/EscapeAnalysis.cpp (hunk 1 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/EscapeAnalysis.cpp:1089]: It means the fields of the value type field haven't been "inlined" into the class that contains that field.  I will reword this to make things more clear.

File: runtime/compiler/optimizer/EscapeAnalysis.cpp (hunk 3 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/EscapeAnalysis.cpp:1197]: Thanks for catching that mistake.

File: runtime/oti/VMHelpers.hpp [hunk 5]
```diff
@@ -756,7 +756,7 @@ class VM_VMHelpers
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 		if (J9_IS_J9CLASS_FLATTENED(arrayClass)) {
 			instance = objectAllocate->inlineAllocateIndexableValueTypeObject(currentThread, arrayClass, size, initializeSlots, memoryBarrier, sizeCheck);
-		} else
+		} else if (J9_ARE_NO_BITS_SET(arrayClass->classFlags, J9ClassContainsUnflattenedFlattenables))
```
Comment (DanHeidinga) [inline runtime/oti/VMHelpers.hpp:759]: I'd like to see TraceAssert added to `inlineAllocateIndexableObject` and similar helpers to validate that we're not passing any classes that have the `J9ClassContainsUnflattenedFlattenables` bit set.

The assert should be guarded by `J9VM_OPT_VALHALLA_VALUE_TYPES`.  FYI @tajila

Comment (DanHeidinga) [reply runtime/oti/VMHelpers.hpp:759]: Looking at the interpreter code, I think we have different semantics there and may trip up on a case like:
```java
inline class I {
   int a;
   int b;
}

class Bad {
  I i;
  int c;
}

new Bad[10];
```
See 
https://github.com/eclipse/openj9/blob/c6b594b6c03f957bb6a07f5d234a6d996037a17f/runtime/vm/BytecodeInterpreter.hpp#L1079-L1087
which calls:
https://github.com/eclipse/openj9/blob/c6b594b6c03f957bb6a07f5d234a6d996037a17f/runtime/vm/BytecodeInterpreter.hpp#L1051-L1061

which means either the inline allocate code already handles this or the interpreter needs to be fixed too

**Comment by tajila** [comment]
@DanHeidinga In the interpreter we have split up the general general array allocation (inline + outOfLine) and inline only allocation. These are `allocateIndexableObject` and `inlineArrayAllocation` respectively. 

Any type performing the inline array allocation must first be tested to see that it doesnt contain the `J9ClassContainsUnflattenedFlattenables`, before performing the inline allocation. This is how it works in `anewarray` bytecode and `allocateIndexableObject` in the interpreter. 

The code in `VM_VMHelpers::allocateIndexableObject` was not updated, so it attempts to do the inline allocation then performs the outofLine. The code in VMHelpers is not structured the same way as the interpreter, the inline allocations are not in a separate function but are bundled together with the outofline allocation so there is a little more flexibility on how to add the check. For arrays, if the `flattened` flag is set, the `J9ClassContainsUnflattenedFlattenables` flag cannot be set. So the proposed solution does work and effectively achieves the same thing as what the interpreter does.

The interpreter does:
1) if `J9ClassContainsUnflattenedFlattenables` perform outofline
2) if `flattened` perform inlineValuetypeArray...
3) else perform inlineArray...
4) if inline fails perform outofline

The proposed VMHelpers do:
1) if `flattened` perform inlineValuetypeArray...
2) if `J9ClassContainsUnflattenedFlattenables` perform outofline
3) else perform inlineArray...
4) if inline fails perform outofline

The order of 1) and 2)  are reversed in comparison to the interpreter code. 

Personally, I think we have too many ways of allocating arrays, which is why this problem arose in the first place. I was thinking about consolidating of them into a single helper, but that should not delay this PR.

File: runtime/oti/VMHelpers.hpp (hunk 5 above)
Comment (DanHeidinga) [reply runtime/oti/VMHelpers.hpp:759]: Based on @tajila's response, the code I missed is in the `anewarray` implementation:
https://github.com/eclipse/openj9/blob/c6b594b6c03f957bb6a07f5d234a6d996037a17f/runtime/vm/BytecodeInterpreter.hpp#L7569-L7574

File: runtime/compiler/compile/J9Compilation.cpp (hunk 4 above)
Comment (hzongaro) [reply runtime/compiler/compile/J9Compilation.cpp]: Good point!  Added a new debug counter in this revised commit:  https://github.com/eclipse/openj9/pull/9139/commits/b27c9e6af6c7645dea2b198333184faa14b3df04

File: runtime/compiler/optimizer/EscapeAnalysis.cpp (hunk 3 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/EscapeAnalysis.cpp:1198]: Corrected in commit https://github.com/eclipse/openj9/pull/9139/commits/d897b437a02626a45d5be525d10a50d192db14c1

File: runtime/compiler/compile/J9Compilation.cpp (hunk 2 above)
Comment (hzongaro) [reply runtime/compiler/compile/J9Compilation.cpp]: Added `const` in commit https://github.com/eclipse/openj9/pull/9139/commits/b27c9e6af6c7645dea2b198333184faa14b3df04

File: runtime/compiler/optimizer/EscapeAnalysis.cpp (hunk 1 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/EscapeAnalysis.cpp]: Revised wording in commit https://github.com/eclipse/openj9/pull/9139/commits/d897b437a02626a45d5be525d10a50d192db14c1

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk14

**Comment by tajila** [comment]
you can run the valuetype tests with `Jenkins test sanity,extended xlinuxval jdknext`

File: runtime/compiler/compile/J9Compilation.cpp [hunk 6]
```diff
@@ -665,11 +667,26 @@ J9::Compilation::canAllocateInline(TR::Node* node, TR_OpaqueClassBlock* &classIn
...
+      // unless value types are enabled, in which case return -1 to prevent inline allocation
       if (classRef->getOpCodeValue() != TR::loadaddr)
          {
          classInfo = NULL;
-         return 0;
+         if (areValueTypesEnabled)
+            {
+            if (self()->getOption(TR_TraceCG))
+               {
+               traceMsg(self(), "cannot inline array allocation @ node %p because value types are enabled\n", node, size);
```
Comment (hzongaro) [inline runtime/compiler/compile/J9Compilation.cpp:679]: Windows build failure due to the reference to the uninitialized `size` variable in the call to `traceMsg`.

Comment (hzongaro) [reply runtime/compiler/compile/J9Compilation.cpp]: Corrected in commit https://github.com/eclipse/openj9/pull/9139/commits/a328c0329eb9ecb0ba1648741437e5cb24d151ad

**Comment by andrewcraik** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by andrewcraik** [comment]
valuetypes test passed

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk14

**Comment by andrewcraik** [comment]
AIX build fail was infra - everything else looks good and the code is not enabled by default so with the build clean this should be good to go.
